### PR TITLE
Disallow nil bindings in VALUES clauses

### PR DESCRIPTION
**Pull request by scottlowe**
Resolves issue #111

### CODE REVIEW

**Comment by scottlowe** [comment]
Relates to private PR at https://github.com/Swirrl/zib/pull/264

File: test/grafter/rdf/sparql_test.clj [hunk 1]
```diff
@@ -69,7 +82,26 @@
...
+               "?s ?p ?o ."
+               "}")
+          (#'sparql/rewrite-values-clauses q4 { [:s :p] ::sparql/undef})))
+
+    (is (same-query?
+          (str "SELECT * WHERE {"
+               "VALUES (?s ?p) { (UNDEF UNDEF) }"
+               "?s ?p ?o ."
+               "}")
+          (#'sparql/rewrite-values-clauses q4 {[:s :p] [::sparql/undef]})))
```
Comment (RickMoynihan) [inline test/grafter/rdf/sparql_test.clj]: The shape of the data in this test seems wrong to me, as the values need to be a list of `[:s :p]` pairs, I think this test should be `[[::sparql/undef ::sparql/undef]]`

Likewise I think we need a test for `[[::sparql/undef (URI. "http://p1")] [(URI. "http://s1") ::sparql/undef]]`.

File: test/grafter/rdf/sparql_test.clj [hunk 2]
```diff
@@ -69,7 +82,26 @@
...
               "}")
          (#'sparql/rewrite-values-clauses q4 { [:s :p] [[(URI. "http://s1") (URI. "http://p1")]
-                                                      [(URI. "http://s2") (URI. "http://p2")]]}))))
+                                                      [(URI. "http://s2") (URI. "http://p2")]]})))
+    (is (same-query?
+          (str "SELECT * WHERE {"
+               "VALUES (?s ?p) { (UNDEF UNDEF) }"
+               "?s ?p ?o ."
+               "}")
+          (#'sparql/rewrite-values-clauses q4 { [:s :p] ::sparql/undef})))
```
Comment (RickMoynihan) [inline test/grafter/rdf/sparql_test.clj]: I've had to think about this one a bit, and @RicSwirrl helped me clarify the thinking on it.  But I think we should not allow use of `::sparql/undef` like this, as it's semantically at the wrong level of nesting.  

I think the intention here though is good, which is to allow the setting of a default behaviour for when you want to override an existing default binding.  

Better behaviours would be:

If you have a query of:

`SELECT * WHERE { VALUES ?s { UNDEF } ?s ?p ?o }`

Which is semantically equivalent to `select ?s ?p ?o ...`

Then providing a binding of `{:s []}` would effectively rewrite to `SELECT * WHERE { VALUES ?s {} ?s ?p ?o }`.  Likewise a `nil` `:s` should do the same (though the first expresses the intent more clearly).

If you want the default behaviour (specified in the query) then you simply don't provide an `:s` in your bindings.

Comment (scottlowe) [reply test/grafter/rdf/sparql_test.clj]: I'm confused. In your example above you seem to suggest that we would like to provide a binding of `{:s []}`

Should result in:

```sparql
SELECT * WHERE { 
  VALUES ?s {}
  ?s ?p ?o 
}
```
But the above is not allowed, it must be 

```sparql
SELECT * WHERE { 
  VALUES ?s { UNDEF }
  ?s ?p ?o 
}
```
... which is what we currently have.

Comment (RickMoynihan) [reply test/grafter/rdf/sparql_test.clj]: The syntax should be supported by stardog (at least 5.x) and definitely the SPARQL standard.  It's the equivalent of a sub select returning no results.

**Pull request merged**